### PR TITLE
Added cooldown for COMP claims

### PR DESCRIPTION
**Pull request by Pet3ris**
**WORK IN PROGRESS**

**Authors**:

- @pluriholonomic, @jmomx, @reichiang, @Pet3ris

**High-level motivation:**

https://www.notion.so/PUBLIC-Improving-2-step-cool-down-0ede3a15fe5f4e4a87129bcb75b9a156

**Key features**:

- `claimComp` now triggers a cooldown period, after which `claimComp` can be called again to claim COMP
- This type of `claimComp` will set `compLocked[user] = compAccrued[user]` and `lastCooldownBlock[user] = block.number`
- On any claim, at most `compLocked[user]` COMP will be claimable and only at time where `block.number >= lastCooldownBlock[user] + cooldownPeriod`
- Added `_setCooldownPeriod` for governance to manage cooldown period

**Specification matrix**

This matrix explains in more detail how different conditions are handled.

|Condition|Behavior if cooldown period = 0|Behavior if cooldown period = X|
|---|---|---|
|First `claimComp`|Claims available amount|Doesn't claim anything, starts cool down timer with all COMP in `compAccrued[user]`|
|`claimComp` after cool down but before cool down period expires|n/a|Does not do anything because cool down timer in effect|
|`claimComp` after cool down period has expired|n/a|Claims COMP that is in `compLocked[user]`, puts all COMP in `compAccrued[user]` in a new cool down cycle and resets timer|

**Testnet testing plan**

Cooldown:
L1 - Cooldown can be changed by admin
L2 - Comptroller works the same way when cooldown = 0
L3 - Cooldown can be initiated on a fresh account
L4 - Cooldown cannot be updated before expiry
L5 - Cooldown can be updated after expiry but before claiming

Updated cooldown claims:
M1 - Nothing is claimed if no tokens locked
M2 - Nothing is claimed if cooldown period hasn't expired
M3 - Full amount is claimed if cooldown period expired and claim amount <= locked amount
M4 - Only locked amount is claimed if cooldown period expired and claim amount > locked amount

General Comptroller consistency:
C1 - Check COMP speed consistency pre/post patch for markets
C2 - Check mint works correctly
C3 - Check redeem works correctly
C4 - Check borrow works correctly
C5 - Check repay works correctly
C6 - Check liquidate works correctly

**Remaining steps**:

- [x] Rebase this change once https://github.com/compound-finance/compound-protocol/pull/83 gets merged
- [x] Finish unit tests
- [x] Finish scenario tests
- [x] Complete audit
- [x] Complete re-audit
- [ ] Compound team review
- [ ] Deployment scenarios
- [ ] Deploy to testnet
- [ ] Complete testnet testing
- [ ] Deploy to mainnet
- [ ] Submit proposal
- [ ] Finalise PR if proposal accepted and remove draft label

### CODE REVIEW

File: contracts/ComptrollerG6.sol [hunk 1]
```diff
@@ -0,0 +1,1414 @@
...
+import "./ComptrollerInterface.sol";
+import "./ComptrollerStorage.sol";
+import "./Unitroller.sol";
+import "./Governance/Comp.sol";
+
+/**
+ * @title Compound's Comptroller Contract
+ * @author Compound
+ */
+contract ComptrollerG6 is ComptrollerV5Storage, ComptrollerInterface, ComptrollerErrorReporter, ExponentialNoError {
```
Comment (coburncoburn) [inline contracts/ComptrollerG6.sol:15]: why a 6 and 7? maybe needs a rebase

Comment (Pet3ris) [reply contracts/ComptrollerG6.sol:15]: @coburncoburn - absolutely, this still has the manual claims patch in there!

Comment (Pet3ris) [reply contracts/ComptrollerG6.sol]: @coburncoburn fixed now.

**Comment by byrmarkn** [comment]
40333312

**Comment by Pet3ris** [comment]
Archiving this branch to preserve audit reference commits.

**Pull request closed without merging**